### PR TITLE
Create permissions utils

### PR DESCRIPTION
**Pull request by OmerLevy1248**

### CODE REVIEW

File: app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    /**
+     * Ask the user for permission.
+     */
+    public static void requestPermissions(final Activity activity) {
+
+        // Request permissions from the user
+        boolean shouldProvideRationale =
+                ActivityCompat.shouldShowRequestPermissionRationale(activity,
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java:48]: Why did you add it to specific permission ?
it's easy to change it to generic with permission name?

File: app/src/main/java/com/gusta/wakemehome/utilities/NotificationUtils.java [hunk 2]
```diff
@@ -8,24 +9,117 @@
...
+     * Show a text message to the user.
+     *
+     * @param activity           The activity to show the message on.
+     * @param mainTextStringId   The id for the string resource for the message's main text.
+     * @param actionTextStringId The id for the string resource for the text of the action item.
+     * @param listener           The listener associated with the Snackbar action.
+     */
+    public static void notifyUser(Activity activity, int mainTextStringId, int actionTextStringId,
+                                  View.OnClickListener listener) {
+        notifyUser(activity, activity.getString(mainTextStringId), null,
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/utilities/NotificationUtils.java:47]: which activity will be send on service?

File: app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java [hunk 3]
```diff
@@ -77,7 +71,8 @@
     // PUBLIC METHODS //
     //================//
 
-    public GeofenceManager(ContextWrapper activity, LiveData<? extends List<? extends GeofenceEntry>> liveData) {
+    public GeofenceManager(ContextWrapper activity,
+                           LiveData<? extends List<? extends GeofenceEntry>> liveData) {
         mContextWrapper = activity;
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java:76]: same question about the activity..

File: app/src/main/java/com/gusta/wakemehome/AlarmAdapter.java [hunk 4]
```diff
@@ -80,7 +80,7 @@ public void onBindViewHolder(@NonNull AlarmViewHolder holder, int position) {
         final AlarmEntry alarmEntry = mAlarmEntries.get(position);
         String location = alarmEntry.getLocation();
         String radius =
-                WakeMeHomeUnitsUtils.formatLength(mContext, alarmEntry.getRadius()) + " radius";
+                UnitsUtils.formatLength(mContext, alarmEntry.getRadius()) + " radius";
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/AlarmAdapter.java]: radius should set to const

**Review by Yuvalkeidar — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java [hunk 5]
```diff
@@ -276,102 +277,15 @@ private void handleError(Exception e) {
...
-        // Request permissions from the user
-        final Activity activity = (Activity) mContextWrapper;
-        boolean shouldProvideRationale =
-                ActivityCompat.shouldShowRequestPermissionRationale(activity,
-                        Manifest.permission.ACCESS_FINE_LOCATION);
-
-        // Provide an additional rationale to the user. This would happen if the user denied the
-        // request previously, but didn't check the "Don't ask again" checkbox.
-        if (shouldProvideRationale) {
-            Log.i(TAG, "Displaying permission rationale to provide additional context.");
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java:349]: I think it should be added to `strings.xml`

File: app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java [hunk 6]
```diff
@@ -276,102 +277,15 @@ private void handleError(Exception e) {
...
-                        @Override
-                        public void onClick(View view) {
-                            // Request permission
-                            ActivityCompat.requestPermissions(activity,
-                                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
-                                    Constants.ACCESS_FINE_LOCATION_PERMISSION_REQUEST_CODE);
-                        }
-                    });
-        } else {
-            Log.i(TAG, "Requesting permission");
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java:361]: `strings.xml`

File: app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                        Manifest.permission.ACCESS_FINE_LOCATION);
+
+        // Provide an additional rationale to the user. This would happen if the user denied the
+        // request previously, but didn't check the "Don't ask again" checkbox.
+        if (shouldProvideRationale) {
+            Log.i(TAG, "Displaying permission rationale to provide additional context.");
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java:54]: `Strings.xml`

File: app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                        @Override
+                        public void onClick(View view) {
+                            // Request permission
+                            ActivityCompat.requestPermissions(activity,
+                                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
+                                    Constants.ACCESS_FINE_LOCATION_PERMISSION_REQUEST_CODE);
+                        }
+                    });
+        } else {
+            Log.i(TAG, "Requesting permission");
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java:66]: `String.xml`

File: app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * Callback received when a permissions request has been completed. Redirected from activity.
+     */
+    public static void handleResult(final Activity activity, PendingTaskHandler pendingTaskHandler,
+                                    @NonNull int[] grantResults) {
+        if (grantResults.length <= 0) {
+            // If user interaction was interrupted, the permission request is cancelled and you
+            // receive empty arrays.
+            Log.i(TAG, "User interaction was cancelled.");
```
Comment (Yuvalkeidar) [inline app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java:84]: All the string into `Strings.xml`, it should be able to be translatable

**Review by Yuvalkeidar — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java (hunk 1 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java:48]: You are right I will change.
I thought it would be problematic for the "settings" link in the snackbar, but it won't (all permissions should send user to "app details" setting).

File: app/src/main/java/com/gusta/wakemehome/utilities/NotificationUtils.java (hunk 2 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/utilities/NotificationUtils.java:47]: A service can't ask for permissions as it can't show the dialog box (actually it can but it's bad user experience)

File: app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java (hunk 3 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java:76]: same answer :)

File: app/src/main/java/com/gusta/wakemehome/AlarmAdapter.java (hunk 4 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/AlarmAdapter.java]: What do you mean?

File: app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java (hunk 5 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java:349]: This is a debug log - if we start adding them to "strings.xml" it will just make it a huge file we can't find anything at (Plus, while debugging we would usually add/removes logs like that - it will just irritate us to put them in strings.xml)

File: app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java (hunk 6 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/geofencing/GeofenceManager.java:361]: Debug log

File: app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java (hunk 7 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java:54]: Debug log

File: app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java (hunk 8 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java:66]: Debug log

File: app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java (hunk 9 above)
Comment (OmerLevy1248) [reply app/src/main/java/com/gusta/wakemehome/utilities/PermissionUtils.java:84]: Debug log.
No need to translate log. Only things that user sees.

**Review by Yuvalkeidar — APPROVED** [review]
